interchange/DeviceResources.capnp

```diff
@@ -536,29 +536,57 @@ struct Device {
   ######################################
   struct LutDefinitions {
     struct LutCell {
-      cell      @0 : Text;
-      inputPins @1 : List(Text);
+      # What cell type is this?
+      cell        @0 : Text;
+
+      # What pins are part of the LUT equations?
+      # Pins are listed in LSB first order.
+      inputPins   @1 : List(Text);
+
+      # How is the LUT equation stored?
+      equation : union {
+        # Equation is stored as an INIT style parameter.
+        # For a LUT2, INIT is 4 bits.
+        #   INIT[0] is output when all pins are 0.
+        #   INIT[1] is output when first pin is 1 and all other pins are 0.
+        #   INIT[2] is output when second pin is 1 and all other pins are 0.
+        #   INIT[3] is output when both pins are 1.
+        initParam @2 : Text;
```

Comment:
LUT INIT strings for Xilinx designs seem to use a Verilog-style hex encoding.  Is that what is assumed here or at least supported?  There is an equation syntax, but I don't believe it is stored as an INIT value.

Reply:
Ya, basically the `initParam` is a verilog-constant that is the INIT value.  Like you say, it could be a Verilog hex, or Verilog bit, etc.

Reply:
The Lattice Nexus parts do the _horrible_ thing of a C-style hex encoding inside a Verilog string: https://github.com/YosysHQ/yosys/blob/master/techlibs/nexus/cells_sim.v#L3 but for the purposes of this I'd be happy to use another primitive with a more logical encoding scheme.

Another special case for the Nexus and ECP5 are the carry LUTs (CCU2s) which are essentially fractured LUT4s with a bottommost LUT2/3 (depending on device) 'tap' feeding carry-related logic but not generally accessible (this _might_ be a bit similar to the Versal carry structure but I haven't looked in detail at that).

Reply:
My suggestion is when the time comes to add a new field to the union like "cHexInitParam", etc.  The whole reason I put a union here is to accomidate future `Text`/`struct` fields to handle arches that require additional exceptions.  Does that make sense?

Reply:
> Another special case for the Nexus and ECP5 are the carry LUTs (CCU2s) which are essentially fractured LUT4s with a bottommost LUT2/3 (depending on device) 'tap' feeding carry-related logic but not generally accessible (this _might_ be a bit similar to the Versal carry structure but I haven't looked in detail at that).

If I understand what you are describing, that sounds sort of like the `LUT6_2` in 7-series/US+.  From an FPGA interchange standpoint, the `LUT6_2` is a macro that expands to a `LUT6` and `LUT5`, that neatly maps onto inputs -> output model presented here.  Until I get around to actually implementing macro expansion "here be dragons", but I believe this can be handled.  Thoughts?